src/main/scala/io/iohk/ethereum/network/discovery/PeerDiscoveryManager.scala

```diff
@@ -145,34 +180,73 @@ class PeerDiscoveryManager(
   def sendDiscoveredNodesInfo(
       maybeDiscoveryService: Option[v4.DiscoveryService],
       recipient: ActorRef
-  ): Unit = {
+  ): Unit = pipeToRecipient(recipient) {
 
     val maybeDiscoveredNodes: Task[Set[Node]] =
       maybeDiscoveryService.fold(Task.pure(Set.empty[Node])) {
         _.getNodes.map { nodes =>
-          nodes.map { node =>
-            Node(
-              id = ByteString(node.id.toByteArray),
-              addr = node.address.ip,
-              tcpPort = node.address.tcpPort,
-              udpPort = node.address.udpPort
-            )
-          }
+          nodes.map(toNode)
         }
       }
 
     maybeDiscoveredNodes
       .map(_ ++ alreadyDiscoveredNodes)
-      .map(_.filterNot(_.id == localNodeId))
+      .map(_.filterNot(isLocalNode))
       .map(DiscoveredNodesInfo(_))
+  }
+
+  def sendRandomNodeInfo(
+      maybeRandomNodes: Option[RandomNodes],
+      recipient: ActorRef
+  ): Unit = pipeToRecipient[RandomNodeInfo](recipient) {
+    val randomNode = maybeRandomNodes match {
+      case None if alreadyDiscoveredNodes.isEmpty =>
+        Task.raiseError(
+          new IllegalStateException("There are no known nodes and discovery isn't running.")
+        )
+
+      case None =>
+        Task.pure(alreadyDiscoveredNodes(Random.nextInt(alreadyDiscoveredNodes.size)))
```

Comment:
Shouldn't the PeerManagerActor most likely already have received this node when he received `PeerDiscoveryManager.DiscoveredNodesInfo`? Or is the purpose of this that he attempts to re-connect?

Reply:
You are correct, if it has asked for `DiscoveredNodesInfo` it will have already got this node, as well as everything else. However we know that when it does so, it only ever takes up to 45 nodes from it, tries to connect, and only consider the some of the rest when it tries to connect once again. A connection may be blacklisted for just 6 minutes, before attempted again, but the `PeerManagerActor` doesn't do that on its own AFAIK, so this could be a way of trying again. Put it another way: it wouldn't ask if it doesn't need one. 

Also here we cannot be certain that it's the `PeerManagerActor` asking for a random node. 

In general a random node is also not unlikely to have been seen before in the case when we do random `lookup`s, since the same nodes can be part of different search results.